File: modules/cloudwatch-metrics/src/main/clojure/xtdb/metrics/cloudwatch.clj
(ns xtdb.metrics.cloudwatch
  (:require [xtdb.metrics :as metrics])
  (:import (io.micrometer.cloudwatch2 CloudWatchConfig CloudWatchMeterRegistry)
           (io.micrometer.core.instrument Clock MeterRegistry)
           (software.amazon.awssdk.services.cloudwatch CloudWatchAsyncClient)
           (xtdb.metrics CloudWatch CloudWatch$Factory)))

(defmethod metrics/->meter-registry-factory ::meter-registry [_ {:keys [namespace ^CloudWatchAsyncClient client]
                                                                 :or {namespace "xtdb.metrics"
                                                                      ^CloudWatchAsyncClient client (CloudWatchAsyncClient/create)}}]
  (CloudWatch/cloudWatch namespace client))

(defn create-meter-registry ^MeterRegistry [^CloudWatch$Factory factory]
  (CloudWatchMeterRegistry.
   (reify CloudWatchConfig
     (get [_ _s] nil)
     (namespace [_] (.getNamespace factory)))
   Clock/SYSTEM
   (.getClient factory)))

File: modules/cloudwatch-metrics/src/main/kotlin/xtdb/metrics/CloudWatch.kt
@file:UseSerializers(PathWithEnvVarSerde::class, StringWithEnvVarSerde::class)

package xtdb.metrics

import io.micrometer.core.instrument.MeterRegistry
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import kotlinx.serialization.UseSerializers
import xtdb.api.PathWithEnvVarSerde
import xtdb.api.StringWithEnvVarSerde
import  software.amazon.awssdk.services.cloudwatch.CloudWatchAsyncClient
import xtdb.api.module.XtdbModule
import xtdb.util.requiringResolve

object CloudWatch {
    @JvmStatic
    fun cloudWatch(namespace: String) = Factory(namespace)

    @JvmStatic
    fun cloudWatch(namespace: String, client: CloudWatchAsyncClient) = Factory(namespace, client)

    @Serializable
    @SerialName("!CloudWatch")
    data class Factory(
        @Serializable(StringWithEnvVarSerde::class) val namespace: String,
        @Transient var client: CloudWatchAsyncClient = CloudWatchAsyncClient.create(),
    ) : MeterRegistryFactory {

        override fun create() = requiringResolve("xtdb.metrics.cloudwatch/create-meter-registry")(this) as MeterRegistry
    }

    /**
     * @suppress
     */
    class Registration : XtdbModule.Registration {
        override fun register(registry: XtdbModule.Registry) {
            registry.registerMeterRegistry(Factory::class)
        }
    }
}